File: backend/app/schemas/evidence.py
"""Pydantic v2 schemas for EvidenceArtifact."""

from datetime import datetime
from uuid import UUID
from typing import Optional

from pydantic import BaseModel, ConfigDict

from app.db.models.enums import ArtifactType


class EvidenceArtifactBase(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    trip_id: UUID
    artifact_type: ArtifactType
    s3_key: str
    s3_bucket: str
    file_hash: str
    mime_type: str
    captured_at: datetime
    captured_by_driver_id: Optional[UUID] = None
    captured_by_user_id: Optional[UUID] = None
    captured_lat: Optional[float] = None
    captured_lng: Optional[float] = None


class EvidenceArtifactCreate(EvidenceArtifactBase):
    pass


class EvidenceArtifactUpdate(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    # Artifacts are immutable after creation; only metadata fields may be patched.
    captured_lat: Optional[float] = None
    captured_lng: Optional[float] = None


class EvidenceArtifactRead(EvidenceArtifactBase):
    id: UUID
    created_at: datetime


class EvidenceArtifactWithUrl(EvidenceArtifactRead):
    """Dispatcher read shape: metadata plus a short-lived signed URL.

    A subclass, not a field on EvidenceArtifactRead, since that's the driver PWA's POST
    response and drivers have no business receiving read URLs. signed_url is None when
    Storage declined to sign — the row is still evidence, just image-unavailable.
    """

    signed_url: Optional[str] = None

File: frontend/shared/lib/types/evidence.ts
// EvidenceArtifact: a photo or document captured during a handshake or checkpoint.
// Stored in Supabase Storage; only the SHA-256 hash reaches the blockchain.
// Mirrors backend EvidenceArtifactRead schema in schemas/evidence.py.

export type ArtifactId = string & { readonly __brand: 'ArtifactId' }

export type ArtifactType = 'photo' | 'document'

export interface EvidenceArtifact {
  id: ArtifactId
  trip_id: string
  artifact_type: ArtifactType
  s3_key: string
  s3_bucket: string
  file_hash: string
  mime_type: string
  captured_at: string
  captured_by_driver_id: string | null
  captured_by_user_id: string | null
  captured_lat: number | null
  captured_lng: number | null
  created_at: string
}

// Dispatcher read shape — metadata plus a short-lived signed Storage URL.
// Mirrors backend EvidenceArtifactWithUrl. `signed_url` is null when Storage declined to
// sign: the artifact is still evidence and its hash still stands, so render the record
// with the image unavailable rather than hiding it.
export interface EvidenceArtifactWithUrl extends EvidenceArtifact {
  signed_url: string | null
}
